### ISSUE TITLE
when there is an error, `proselint --json` should output the error to `stderr` rather than `stdout`

### ISSUE THREAD

**Issue by milanglacier** [amperser/proselint#1307]
I am using `proselint` with `null-ls` which is a neovim plugin, essentially what this plugin does is `cat xxx.md | proselint --json` (assuming I am editing a markdown file),
sometimes what proseling output will be not a valid json, 
i.e something like this:
```
output: Error ([Errno 35] Resource temporarily unavailable: '/Users/northyear/.cache/proselint/checks.airlinese.misc.check') opening /Users/northyear/.cache/proselint/checks.airlinese.misc.check - will attempt to delete and re-open.

```
which may from here,
https://github.com/amperser/proselint/blob/ea8c5749d4ea4fe1a137e45885254e71cbcefa98/proselint/tools.py#L65

i.e from this line. Since it is not a legal json,  and then `null-ls` will parse it with error.
It would be better that when `--json` is supplied, errors should be outputed to `stderr` rather than `stdout`.

**Comment by jose-elias-alvarez** [comment]
To provide a little additional detail here: null-ls is a plugin that spawns external processes like `proselint`, captures their output, and converts that output into Neovim's diagnostic format for display in the editor. In this case, the issue stems from the error message, which is sent to `stdout` and therefore breaks the plugin's attempts to parse it as JSON. 

Ideally, running `proselint` with the `--json` flag would always output valid JSON. I'm not a `proselint` user myself, so I haven't done research into other integrations, but I imagine they'd run into the same problem if they run under the conditions that create the error message. A simple solution (and one that I've seen other linters use) is to output error messages like these to `stderr`, which will fix things on our end while still providing feedback to users of the CLI. I'm not a Python expert, but I'm happy to put in a PR if there's a chance of getting this fixed.

**Comment by Nytelife26** [comment]
> To provide a little additional detail here: null-ls is a plugin that spawns external processes like `proselint`, captures their output, and converts that output into Neovim's diagnostic format for display in the editor. In this case, the issue stems from the error message, which is sent to `stdout` and therefore breaks the plugin's attempts to parse it as JSON. 

I use null-ls with Neovim on a daily basis, but thank you. It's an honour to speak with you.

This is strange behaviour on part of `proselint` indeed. I picked up the project a while ago, so I haven't yet squashed all the bugs. As you and @milanglacier have suggested, I'll look into why `proselint` sometimes fails to produce valid JSON, and also consider throwing to `stderr`.

**Comment by milanglacier** [comment]
I find that if proselint is installed by the python installed by homebrew, then it will continuously give me such warning, i.e it can not open the file in ~/.cache/proselint folder.
However, if I install proselint via a venv in my home and not use the python installed by homebrew (in fact the python from miniforge) and then it works fine and no warning.

**Comment by Nytelife26** [comment]
> It seems that `[Errno 35] Resource temporarily unavailable` is usually a macOS only problem in many scenarios😂

Odd, I suspected it would be a filesystem issue persistent with most modern UNIX systems. Alas, I should expect nothing less from MacOS 😂 I'll investigate later tonight. Fortunately we have MacOS available in GitHub actions so I can test it that way.

**Comment by mcepl** [comment]
Actually, it doesn’t seem to be reproducible on Linux as well.

Using neovim nightly (commit 62916946), proselint 0.13.0, null-ls actual).

**Comment by mabuchner** [comment]
> I find that if proselint is installed by the python installed by homebrew, then it will continuously give me such warning, i.e it can not open the file in ~/.cache/proselint folder.

I think, it's a permission issue. On my MacOS system the permissions for `~/.cache/proselint` were set to `drwx--x--x`.

After changing the permissions

```
chmod g+r,o+r proselint
```

the error went away.

Of course, it would still be nice, if `proselint` would output the error message to `stderr`.

**Comment by Nytelife26** [comment]
I have updates for you all, at long last.

As of #1400, proselint outputs to stderr in JSON mode, although not specifically when there are problems. I will have to rectify that behavior in future. The invalid JSON issue appears to be the result of cache failures, which would be resolved in #1383 when the cache was removed.

**Comment by jclicodes** [comment]
For anyone reading this thread: I have managed to get a successful configuration with Neovim's none-ls and proselint.

First, install proselint using the latest commit:
`pipx install git+https://github.com/amperser/proselint.git`

I'm using `pipx` but you can do this with `pip` too. Make sure to specify --user flag if your python distribution is externally managed by your OS. e.g. `python3 -m pip install --user git+https://github.com/amperser/proselint.git`

Now test proselint to make sure its working. The syntax of the latest commit is different from the latest release. Use the below commands to test:

```
proselint check -o json < path/to/your/text/file 1>/tmp/proselint.out 2>/tmp/proselint.err
cat /tmp/proselint.out
cat /tmp/proselint.err
```

You should observe that there is no output from STDOUT and that there is JSON output in STDERR

Now, onto Neovim:

The latest commit has a different output schema in JSON mode, so we need to register a custom source with Null-ls and do the parsing ourselves. I'm using None-ls here but as far as I know, you can do the same thing with Null-ls.

Register the latest proselint as a source (example using Lazy as a package manager):

```lua
return {
  "nvimtools/none-ls.nvim",
  dependencies = { "nvim-lua/plenary.nvim" },
  config = function()
    local null_ls = require("null-ls")
    local helpers = require("null-ls.helpers")

    local proselint_dev = {
      name = "proselint-dev",
      method = null_ls.methods.DIAGNOSTICS,
      filetypes = { "markdown", "text" },
      generator = null_ls.generator({
        command = "proselint",
        args = { "check", "-o", "json" },
        to_stdin = true,
        from_stderr = true,
        format = "json",
        check_exit_code = function(code)
          return code <= 1
        end,
        on_output = function(params)
          local out = params.output
          if not out or out.status ~= "success" then
            return {}
          end
          local errors = (out.data and out.data.errors) or {}
          local sev = { error = 1, warning = 2, suggestion = 4 }
          local diags = {}

          for _, d in ipairs(errors) do
            local row = tonumber(d.line) or 1
            local start_pos = tonumber(d.start_pos)
            local end_pos = tonumber(d.end_pos)
            local col = tonumber(d.column)

            local start_col, end_col
            if start_pos and end_pos then
              start_col = start_pos
              end_col = math.max(end_pos, start_pos + 1)
            elseif col then
              start_col = col
              end_col = col + 1
            else
              start_col, end_col = 1, 2
            end

            table.insert(diags, {
              row = row,
              col = start_col,
              end_col = end_col,
              message = d.message or "",
              code = d.check_path or "proselint",
              severity = sev[d.severity] or 2,
              source = "proselint",
            })
          end

          return diags
        end,
      }),
    }

    null_ls.register({ proselint_dev })

    -- you can set up your builtin sources as normal here.
    null_ls.setup({ }) 
  end,
}
```

Now open your text file in Neovim, it should be working! Check :NullLsLog and ensure there are no warnings. If there are, you can specify debug = true in the Null-ls setup and see the issue.

**Comment by jclicodes** [comment]
@Nytelife26 FYI, after some more thorough testing I found some issues with my original implementation which I have since fixed (and edited the original message)

proselint stable:
```json
      {
        "check": "needless_variants.misc",
        "column": 69,
        "end": 122,
        "extent": 10,
        "line": 5,
        "message": "Needless variant. 'oversize' is the preferred form.",
        "replacements": "oversize",
        "severity": "warning",
        "start": 112
      },
```

proselint (latest commit):
```
      {
        "check_path": "needless_variants",
        "column": 69,
        "end_pos": 121,
        "line": 5,
        "message": "Needless variant. 'oversize' is the preferred form.",
        "replacements": "oversize",
        "severity": "warning",
        "start_pos": 112
      },
```

1) There is no more 'extent' field to find the end position of the diagnostic

2) Proselint gives the index of the character whereas Neovim expects the index of the byte.

So I had to add some code to perform the necessary conversions and calculate the extent by finding the range of end_pos and start_pos. It now appears to be working properly.

**Comment by Nytelife26** [comment]
Yes, both of those are quite intentional. It is likely that the wire schema will change in other backwards-incompatible ways in the near future too, so I will make an issue outlining the plans for that.

I would appreciate your feedback on those plans, if you'd like to review them, since you have experience using the JSON output for external applications.

**Comment by jclicodes** [comment]
I would be happy to. Ping me when you get around to creating the issue, or email me at jake@jcli.codes
